File: commonlibrary/src/main/java/com/panda/commonlibrary/mvp/BasePresenter.java
package com.panda.commonlibrary.mvp;

import android.content.Context;

import com.panda.commonlibrary.BaseApp;
import com.trello.rxlifecycle3.LifecycleTransformer;
import com.trello.rxlifecycle3.android.ActivityEvent;
import com.trello.rxlifecycle3.android.FragmentEvent;
import com.trello.rxlifecycle3.components.support.RxAppCompatActivity;
import com.trello.rxlifecycle3.components.support.RxDialogFragment;
import com.trello.rxlifecycle3.components.support.RxFragment;

import java.lang.ref.WeakReference;

/**
 * <pre>
 *     Created by ppW
 *     e-mail : wangpanpan05@163.com
 *     time   : 2019/02/25
 *     desc   :
 *     version: 1.0   初始化
 *     params:
 *  <pre>
 */
public class BasePresenter<V extends BaseContract.View> implements BaseContract.Presenter<V> {
    private WeakReference<V> mView;
    private WeakReference<Context> mContext;
    private LifecycleTransformer<Integer> mLifecycleTransformer;

    public BasePresenter () {
    }

    @Override
    public void attachView (V view) {
        mView = new WeakReference<>(view);
        V v = mView.get();
        if (v instanceof RxAppCompatActivity) {
            mContext = new WeakReference<>((RxAppCompatActivity) v);
            mLifecycleTransformer = ((RxAppCompatActivity) v).bindUntilEvent(ActivityEvent.DESTROY);
        } else if (v instanceof RxFragment) {
            mContext = new WeakReference<>(((RxFragment) v).getContext());
            mLifecycleTransformer = ((RxFragment) v).bindUntilEvent(FragmentEvent.DESTROY);
        } else if (v instanceof RxDialogFragment) {
            mContext = new WeakReference<>(((RxDialogFragment) v).getContext());
            mLifecycleTransformer = ((RxDialogFragment) v).bindUntilEvent(FragmentEvent.DESTROY);
        }
    }

    protected LifecycleTransformer getLifecycleTransformer () {
        return mLifecycleTransformer;
    }

    public Context getContext () {
        if (mContext != null) {
            return mContext.get();
        }
        return BaseApp.Companion.getInstance().getApplicationContext();
    }

    @Override
    public void detachView () {
        if (null != mView) {
            mView.clear();
            mView = null;
        }
        if (mContext != null) {
            mContext.clear();
            mContext = null;
        }
    }

    public V getView () {
        return mView.get();
    }
}

File: commonlibrary/src/main/java/com/panda/commonlibrary/mvp/BasePresenter.kt
package com.panda.commonlibrary.mvp

import android.content.Context
import com.panda.commonlibrary.BaseApp.Companion.getInstance
import com.panda.commonlibrary.net.callback.BaseCallBack
import com.panda.commonlibrary.net.model.BaseResponseBody
import com.trello.rxlifecycle3.LifecycleTransformer
import com.trello.rxlifecycle3.android.ActivityEvent
import com.trello.rxlifecycle3.android.FragmentEvent
import com.trello.rxlifecycle3.components.support.RxAppCompatActivity
import com.trello.rxlifecycle3.components.support.RxDialogFragment
import com.trello.rxlifecycle3.components.support.RxFragment
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import java.lang.ref.WeakReference

/**
 * <pre>
 * Created by ppW
 * e-mail : wangpanpan05@163.com
 * time   : 2019/02/25
 * desc   :
 * version: 1.0   初始化
 * params:
 * <pre>
</pre></pre> */
open class BasePresenter<V : BaseContract.View> : BaseContract.Presenter<V> {
    private lateinit var mView: WeakReference<V>
    private var mContext: WeakReference<Context?>? = null

    override fun attachView(view: V) {
        mView = WeakReference(view)
    }

    val context: Context?
        get() = if (mContext != null) {
            mContext!!.get()
        } else getInstance().applicationContext

    override fun detachView() {
        mView.clear()
        if (mContext != null) {
            mContext!!.clear()
            mContext = null
        }
    }

    private fun <L> getLife(): LifecycleTransformer<L> {
        return when (val v = mView.get()) {
            is RxAppCompatActivity -> {
                mContext =
                    WeakReference(v as RxAppCompatActivity?)
                (v as RxAppCompatActivity).bindUntilEvent(ActivityEvent.DESTROY)
            }
            is RxFragment -> {
                mContext =
                    WeakReference((v as RxFragment).context)
                (v as RxFragment).bindUntilEvent(
                    FragmentEvent.DESTROY
                )
            }
            is RxDialogFragment -> {
                mContext =
                    WeakReference((v as RxDialogFragment).context)
                (v as RxDialogFragment).bindUntilEvent(
                    FragmentEvent.DESTROY
                )
            }
            else -> {
                throw Exception("mView只能绑定RxAppCompatActivity,RxFragment,RxDialogFragment三者之一")
            }
        }
    }

    val view: V?
        get() = mView.get()

    fun <T> postData(
        api: Observable<BaseResponseBody<T>>,
        apiSuccess: (T) -> Unit
    ) {
        api.subscribeOn(Schedulers.io())
            .compose(getLife())
            .map {

                return@map it
            }
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(object : BaseCallBack<T>() {
                override fun onSuccess(data: T) {
                    apiSuccess.invoke(data)
                }

                override fun startLoading() {
                    view?.startLoading()
                }

                override fun endLoading() {
                    view?.endLoading()
                }

                override fun onError(errorMsg: String?) {
                    view?.onError(errorMsg)
                }

                override fun onNetError() {
                    view?.netError()
                }
            })

    }
}